### PR TITLE
Fix security transparency issue

### LINKED ISSUE #79749: System.Diagnostics.Metrics.Tests.MetricEventSourceTests.EventSourceFiltersInstruments failing in CI (Outerloop)
Configuration: `net48-windows-Release-x86-Windows.10.Amd64.Client21H1.Open`

```
Unexpected Error event: System.MethodAccessException: Attempt by security transparent method 'System.Diagnostics.Metrics.MetricsEventSource+MetricSpec.Parse(System.String)' to access security critical method 'System.MemoryExtensions.AsSpan(System.String, Int32, Int32)' failed.

Assembly 'System.Diagnostics.DiagnosticSource, Version=8.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' is marked with the AllowPartiallyTrustedCallersAttribute, and uses the level 2 security transparency model.  Level 2 transparency causes all methods in AllowPartiallyTrustedCallers assemblies to become security transparent by default, which may be the cause of this exception.
at System.Diagnostics.Metrics.MetricsEventSource.MetricSpec.Parse(String text) in //src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:line 474
at System.Diagnostics.Metrics.MetricsEventSource.CommandHandler.ParseSpecs(String metricsSpecs) in //src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:line 393
at System.Diagnostics.Metrics.MetricsEventSource.CommandHandler.OnEventCommand(EventCommandEventArgs command) in /_/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:line 363
Expected: True
Actual:   False
  at System.Diagnostics.Metrics.Tests.MetricsEventListener.AssertOnError() in /_/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs:line 1144
   at System.Diagnostics.Metrics.Tests.MetricsEventListener.GetCountEvents(String eventName) in /_/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs:line 1153
   at System.Diagnostics.Metrics.Tests.MetricsEventListener.WaitForEvent(TimeSpan timeout, Int32 numEvents, String eventName) in /_/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs:line 1117
   at System.Diagnostics.Metrics.Tests.MetricEventSourceTests.EventSourceFiltersInstruments() in /_/src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs:line 272
```

https://dev.azure.com/dnceng-public/public/_build/results?buildId=114328&view=ms.vss-test-web.build-test-results-tab&runId=2422252&paneView=debug

### PR DESCRIPTION
**Pull request by noahfalk**
Fixes #79749

[UPDATE]: The new fix is to remove the APTCA attribute and avoid running in partial trust.

[ORIGINAL fix that is no longer being applied]
This is a partial revert of #78729. That change inadvertently introduced a security transparency issue because the newly added AsSpan() call is security critical and can't be called in a security transparent scope.

### CODE REVIEW

**Comment by stephentoub** [comment]
Why is AsSpan security critical, and why doesn't this impact every other netstandard2.0 library we have that uses AsSpan?

**Comment by jkotas** [comment]
We have removed .NET Framework partial trust support long time ago. It is not possible to use any new language features for  .NET Framework partial trust. Details are in https://github.com/dotnet/runtime/issues/24122. 

Why does the test from https://github.com/dotnet/runtime/issues/79749 run in a partial trust sandbox?

**Comment by noahfalk** [comment]
> Why does the test from https://github.com/dotnet/runtime/issues/79749 run in a partial trust sandbox?

Unknown. It isn't an intentional aspect of the tests for this assembly that it should run under partial trust. I assumed that partial trust was simply a test environment that was occasionally used and I had no desire for these tests to generate spurious failures if they were going to be subjected to that.

> why doesn't this impact every other netstandard2.0 library we have that uses AsSpan?

Perhaps because System.Diagnostics.DiagnosticSource has a lingering APTCA attribute and other libraries don't?
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Diagnostics.DiagnosticSource/src/AssemblyInfo.netfx.cs#L6

I am unaware of what scenario requires S.D.DS to maintain the APTCA attribute so I am fine to remove it. It sounds like that would align this assembly with general approach used by the rest of our libraries?

**Comment by stephentoub** [comment]
> Perhaps because System.Diagnostics.DiagnosticSource has a lingering APTCA attribute and other libraries don't?

Got it.  Yeah, it appears to be the only one left.  (We also seem to have one test that sets it, but no other production assemblies.)

> so I am fine to remove it

That sounds like the right solution to me.

**Comment by jkotas** [comment]
> I am unaware of what scenario requires S.D.DS to maintain the APTCA attribute

FWIW, the APTCA attribute was introduced by https://github.com/dotnet/corefx/pull/17076 .

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj [hunk 1]
```diff
@@ -116,7 +116,6 @@ System.Diagnostics.DiagnosticSource</PackageDescription>
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
     <Compile Include="System\Diagnostics\HttpHandlerDiagnosticListener.cs" />
```
Comment (jkotas) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:118]: You should be also able delete `ALLOW_PARTIALLY_TRUSTED_CALLERS` from this file and all code under this ifdef.

**Comment by jkotas** [comment]
Could you please also delete `ALLOW_PARTIALLY_TRUSTED_CALLERS` from `src\libraries\Common\src\System\HexConverter.cs`?

LGTM otherwise.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj (hunk 1 above)
Comment (noahfalk) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:118]: Got rid of the rest of it.

**Comment by noahfalk** [comment]
Sure, got HexConverter.cs as well.

**Review by jkotas — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
It'd still be interesting to know why the tests are running in partial trust. @ViktorHofer, do you know?

**Comment by jkotas** [comment]
I do not think that the tests are running in partial trust. 

.NET Framework validates consistency of the SecurityCritical/SecurityTransparent attributes in APTCA assemblies even when the code is running in full trust. It explains why the test is failing.

**Comment by stephentoub** [comment]
> even when the code is running in full trust

Interesting, I didn't know that. Thanks.

**Pull request merged**